Add encryption support and access privileges

klimeryk opened the pull request:
Hi, @diegomura! Love your library - I've been using it extensively at https://github.com/klimeryk/recalendar.js/ (https://recalendar.me/). I stumbled upon https://github.com/diegomura/react-pdf/issues/672 and figured it was a great chance to contribute back :)

I've tried my best to bring over the changes from the upstream pdfkit version. Encryption was originally added in https://github.com/foliojs/pdfkit/pull/820 there. It was mostly a straightforward update... except changes in `reference.js`. Looks like that class has been fundamentally changed - upstream it uses internally `Buffer`, while in react-pdf, the class extends `Writeable` instead.
I've tried first keeping the changes to `reference.js` minimal (see https://github.com/diegomura/react-pdf/commit/370fadedcc8f91a82394459972bda6b4722bac82). That was a partial success - non-encrypted PDFs look as before (good), you can generate a PDF that requires a password, the password is properly validated... but then the opened PDF is missing content. See this file: [recalendar-password.pdf](https://github.com/diegomura/react-pdf/files/14809024/recalendar-password.pdf) (password: `secret`)

I've then tried just switching to upstream's `reference.js`, especially seeing your intention of doing that from https://github.com/diegomura/react-pdf/issues/2613. See https://github.com/diegomura/react-pdf/commit/494459f478619cf93fc839e93082afc7aa0af428. That looks more promising - the decrypted PDF has the text and almost looks perfect... except, for some reason, the font seems to be missing, so it looks off. See this file: [recalendar-password2.pdf](https://github.com/diegomura/react-pdf/files/14809017/recalendar-password2.pdf) (password: `secret`)
And this applies to both encrypted _and_ non-encrypted generated PDFs. So I'm guessing there's some modification/customization, maybe in some other file that I missed that was done in this version of pdfkit that is not compatible with the upstream version of `reference.js`? Unfortunately, I could not find it. I've tried also going back in history of `reference.js`, but unfortunately these changes seems to be from before it was moved to monorepo and I cannot find the previous repository 😞 
I figured I'd create this PR for you to have a look, maybe something will spark a quick solution or hint from you. If not, I can continue digging 🙇 

### Testing
I'm guessing you have your preferred ways of testing different use cases :) Enabling password protection simply boils down to passing `userPassword` when creating a new instance of `PDFDocument` like so:
```
const ctx = new PDFDocument( {
    // ...
    userPassword: 'secret',
} );
```

To be on the safe side, I've also tried enabling all optional permissions, but that did not make a difference:
```
    permissions: {                                                                                                                                        
        printing: 'highResolution',
        modifying: true,
        copying: true,
        annotating: true,
        fillingForms: true,
        contentAccessibility: true,
        documentAssembly: true,
    },
```

wojtekmaj commented on packages/pdfkit/src/reference.js:
```diff
@@ -1,91 +1,74 @@
-    return this.document._refEnd(this);
+    this.document._refEnd(this);
   }
-
```
Nit: we probably want to keep this space

klimeryk replied:
Makes sense, I missed this during merge - I've restored it 🙇

wojtekmaj commented on packages/pdfkit/src/security.js:
```diff
@@ -5,33 +5,557 @@
-
-  return Buffer.from(byteArray);
-};
+import CryptoJS from 'crypto-js';
```
I wonder how this change affects bundle size and what was the original reason for selective import done like that.

klimeryk replied:
Yeah, I tried digging in revision history, but it was introduced with the selective import in https://github.com/diegomura/react-pdf/pull/1894. Upstream never used the selective import and imported the whole library from the beginning: https://github.com/foliojs/pdfkit/pull/820. I'm guessing it's because the full/upstream `security.js` file uses extensively this library - `CryptoJS.MD5`, `CryptoJS.RC4`, `CryptoJS.lib.WordArray`, `CryptoJS.SHA256`, `CryptoJS.AES`. Whereas the simplified version (not supporting password protection) indeed just used `CryptoJS.MD5`, so the selective import might have made sense for that use case.

For completeness sake (so that the implications of this PR on the bundle size are known):

#### Current `master`
```
646,480 pdfkit.browser.cjs
644,160 pdfkit.browser.js
326,577 pdfkit.browser.min.cjs
325,782 pdfkit.browser.min.js
407,839 pdfkit.cjs
405,488 pdfkit.js
232,858 pdfkit.min.cjs
231,996 pdfkit.min.js
```

#### This branch/PR
```
683,846 pdfkit.browser.cjs
680,853 pdfkit.browser.js
343,476 pdfkit.browser.min.cjs
342,353 pdfkit.browser.min.js
445,176 pdfkit.cjs
442,171 pdfkit.js
249,962 pdfkit.min.cjs
248,604 pdfkit.min.js
```

So, 9% increase for `pdfkit.js` and 7.2% for `pdfkit.min.js`.

wojtekmaj reviewed (commented):
Woah, nice one! Looks good to me.

I wonder if we could have some tests for that. As long as #2712 repairs them 🥲 

Also, changeset is missing - please add!

klimeryk commented:
Thanks for looking at the PR, @wojtekmaj! Sorry for the late follow-up - private life got in the way, but now I'm back and determined to see this PR through. I've addressed your comments and generated the changeset 🙇 

To be extra clear - **I don't think this PR is ready to merge as-is**. At the very least there's an issue with the fonts. There's links to generated PDF files in the description and my attempt at best describing the possible issue, but it might be best shown on screenshots:

#### Production file, correct fonts

<img width="631" alt="Screenshot 2024-06-14 at 12 10 04" src="https://github.com/diegomura/react-pdf/assets/3392497/53dcc302-0843-4729-b24b-1eedaeca7220">

#### PR file, missing fonts
<img width="631" alt="Screenshot 2024-06-14 at 12 10 12" src="https://github.com/diegomura/react-pdf/assets/3392497/29940a32-ffa1-4c5b-a54f-7332f12d9513">

And there are indeed warnings from the Firefox internal PDF previewer:
```
Warning: getFontFileType: Unable to detect correct font file Type/Subtype.
Warning: FormatError: Required "loca" table is not found
```

My guess is that these attachments/resources are not correctly encoded or decoded with the passwords/security logic introduced in this PR. I was hoping I'd get some pointers from @diegomura or anyone else familiar with PDF internals 🙏  Though, if not possible (I understand that we're all busy one way or another), I'll try my best to dig into - it's an interesting problem and I love understanding how things work. Just I might not be as effective without expert help 🙇

klimeryk commented:
PR updated - it should be good for a final/proper review. The font issues have been solved thanks to @florianbepunkt, see https://github.com/diegomura/react-pdf/issues/2924#issuecomment-2453621314. I've also ported some unit tests from `pdfkit`, see https://github.com/diegomura/react-pdf/commit/798995ceabaf35f053a5f5e7645f417573273cad. Only ported the ones that were passing and trivial to move over. More work needs to be done there, since this version of pdfkit is a subset of upstream. But it will be helpful in the efforts in https://github.com/diegomura/react-pdf/issues/2613

kev2480 commented:
I'd love to see this in the next release! Any updates? 🙏

sengitu-florio commented:
Is there any work required to make this ready for release? I'd be happy to utilize this `userPassword`. I'd be happy to contribute if additional work is required.

andreasss55 commented:
Hi, any timeline when this is going to be released?

klimeryk commented:
Looks like password support got just merged: https://github.com/diegomura/react-pdf/pull/3344, so I'll close this PR 🙇

diegomura commented:
Ouch. So sorry @klimeryk , i did not see this. I got claude to work on it and I didn't check open PRs. I'm so sorry again

klimeryk commented:
@diegomura, no worries, totally understand - ultimately, I'm just glad to have the password support in! I'll be sure to test it out when it's released 💖
